Lineage Compatibility and Support (#579)
astronomer/docs: astro/lineage-support-and-compatibility.md

---
sidebar_label: 'Lineage Support and Compatibility'
title: "Astro Lineage Support and Compatibility Reference"
id: lineage-support-and-compatibility

> description: A compatibility reference guide for Astro lineage's Airflow support.

Author: ```suggestion
description: A compatibility reference guide for Astro's lineage support.
```

> sidebar_label: 'Lineage Support and Compatibility'
> title: "Astro Lineage Support and Compatibility Reference"
> id: lineage-support-and-compatibility
> description: A compatibility reference guide for Astro lineage's Airflow support.

Author: ```suggestion
sidebar_label: 'Data Lineage Support and Compatibility'
title: "Data Lineage Support and Compatibility Reference"
id: data-lineage-support-and-compatibility
description: A compatibility reference guide for Astro lineage's Airflow support.
```

---

## Overview

> This document contains a list Airflow operators which are supported with Astro lineage by default.

Reviewer: A list "of" Airflow operators

> By default, all Astro Deployments gather lineage data from Airflow via the [OpenLineage Airflow integration](https://openlineage.io/integration/apache-airflow/). This integration includes full built-in support for several Airflow operators. These operators emit all available lineage data to your lineage observability dashboard without any additional configuration. Task runs that use these operators will appear as nodes in your data lineage graphs.

Reviewer 2: Remove the word `full`.

Reviewer 2: `Task runs that use these operators will appear as nodes in your data lineage graphs.` --> `Task runs that use these operators will appear as nodes in your data lineage graphs which are connected to any input and output Datasets.`

Author: Not sure we want to get on this level of detail in the overview. What about "Task runs that use these operators will appear as job nodes in your data lineage graphs"?

> By default, all Astro Deployments gather lineage data from Airflow via the [OpenLineage Airflow integration](https://openlineage.io/integration/apache-airflow/). This integration includes full built-in support for several Airflow operators. These operators emit all available lineage data to your lineage observability dashboard without any additional configuration. Task runs that use these operators will appear as nodes in your data lineage graphs.

Reviewer 3: Based on @conorbev's clarification in slack

> By default, all Astro Deployments gather lineage data from Airflow via the [OpenLineage Airflow integration](https://openlineage.io/integration/apache-airflow/). This integration includes full built-in support for several Airflow operators. These operators emit all available lineage data to your lineage observability dashboard without any additional configuration. Task runs that use these operators will appear as nodes in your data lineage graphs.

Author: I think the detail about what happens when an operator is NOT supported should go later in the doc.

> Supported operators collect lineage data via monitoring tools called extractors. If you’re using an operator that isn't currently supported for lineage, we recommend either creating an issue in the [OpenLineage GitHub](https://github.com/OpenLineage/OpenLineage) or writing your own custom extractor.

Reviewer 3: ```suggestion
If you’re using an operator that isn't currently supported for lineage, we recommend either creating an issue in the [OpenLineage GitHub repository](https://github.com/OpenLineage/OpenLineage) or writing your own custom extractor.
```

> ## Fully Supported Airflow Operators

Reviewer 3: Is there an Astro Runtime version that's minimum here? 4.2.0?

The following operators are fully supported in Astro lineage:

- `PostgresOperator`
- `BigQueryOperator`

> - `SnowflakeOperator`

Author: If this operator is supported, does that mean you will also see any Snowflake-based DBs appear on the lineage graph? Or just the task? 

Conversely, if you integrate Databricks, will a task that triggers a databricks job emit any lineage data? Or will a databricks job appear as a job completely separate from Airflow until that data comes into contact with one of these listed operators?

Reviewer: In the case of snowflake, it will only show tasks and datasets connected to tasks.

For databricks, the task that calls the underlying Spark stuff will show up as a job, but without connections. The underlying Spark tasks will emit their own lineage events, and if they touch datasets that are also touched by other Airflow tasks it should show a graph connecting them.

> The following operators are fully supported in Astro lineage:

Reviewer 2: We've found some bugs with the SQL parser and run also realized that lineage isn't currently extracted when a DB connection is being stored in the secrets backend. I therefore think we should remove the word `fully` from this sentence.

- `GreatExpectationsOperator`

:::tip

> The `GreatExpectationsOperator` additionally emits data quality information to the lineage UI's **Quality** tab.

Reviewer 3: ```suggestion
The `GreatExpectationsOperator` additionally emits data quality information to **Quality** tab in the **Lineage** view of the Cloud UI. For more information, see [Data Lineage on Astro](data-lineage.md).
```

:::

## Partially Supported Airflow Operators

> The following operators are partially supported in Astro lineage:

Reviewer 3: ```suggestion
The following operators are partially supported by the Airflow integration with Open Lineage:
```

- `PythonOperator`
- `BashOperator`

Partially supported operators:

- Appear in the lineage UI graph view as nodes when used in an Airflow task.

> - Emit source code and task run-based lineage data.

Reviewer 4: It does emit the source code, but that doesn't show up in the UI, so I think we cut that comment and reduce to "Emit task run-based ..."

> Partially supported operators:
> 
> - Appear in the lineage UI graph view as nodes when used in an Airflow task.
> - Emit task run-based lineage data.
> - Do not emit lineage data about input or output datasets.

Reviewer 3: ```suggestion
Airflow tasks that are run with partially supported operators:

- Emit task run-based lineage data.
- Appear in the graph view of the **Lineage** tab in the Cloud UI as nodes.
- Do not emit lineage data about input or output datasets.
```

## Other Known Limitations

Lineage on Astro is still in active development. Keep in mind the following limitations when using lineage functionality:

> - Until Airflow 2.3.0 is released, supported Airflow operators will not emit lineage data about failed task runs.

Reviewer 3: ```suggestion
- Airflow operators will emit lineage data about failed task runs only for Deployments on Astro Runtime v5.0+
```

> - [Add more limitations here]

Reviewer 2: Data Lineage cannot currently be extracted from Airflow Connections stored in the [Secrets Backend](https://docs.astronomer.io/astro/secrets-backend/#overview)

> - [Add more limitations here]

Reviewer 3: ```suggestion
- Data lineage cannot currently be extracted from datasets whose Airflow Connection is stored in a [Secrets Backend](https://docs.astronomer.io/astro/secrets-backend/#overview)
```
